embedded/document/engine.go

```diff
@@ -151,539 +52,780 @@ func NewEngine(store *store.ImmuStore, opts *sql.Options) (*Engine, error) {
+	for i, index := range indexes {
+		fields := make([]string, len(index.Cols()))
+
+		for i, c := range index.Cols() {
+			fields[i] = c.Name()
 		}
 
-		_, _, err := e.sqlEngine.ExecPreparedStmts(
-			ctx,
-			sqlTx,
-			deleteIdxStmts,
-			nil,
-		)
-		if err != nil {
-			return err
+		collection.Indexes[i] = &protomodel.Index{
+			Fields:   fields,
+			IsUnique: index.IsUnique(),
 		}
 	}
 
-	if len(addIdxKeys) > 0 {
-		// add index keys
-		for name, idx := range addIdxKeys {
-			colLen, err := valueTypeDefaultLength(idx.Type)
-			if err != nil {
-				return fmt.Errorf("index key specified is not supported: %v", idx.Type)
-			}
+	return collection
+}
 
-			// check if index column already exists
-			if _, err := table.GetColumnByName(name); err == nil {
-				continue
-			}
+func (e *Engine) UpdateCollection(ctx context.Context, collectionName string, idFieldName string) error {
+	if idFieldName == DocumentBLOBField {
+		return fmt.Errorf("%w(%s)", ErrReservedFieldName, DocumentBLOBField)
+	}
 
-			// add indexes as new columns to collection
-			updateCollectionStmts = append(updateCollectionStmts, sql.NewAddColumnStmt(collectionName, sql.NewColSpec(name, idx.Type, colLen, false, false)))
-		}
+	opts := sql.DefaultTxOptions().
+		WithUnsafeMVCC(true).
+		WithSnapshotMustIncludeTxID(func(lastPrecommittedTxID uint64) uint64 { return 0 }).
+		WithSnapshotRenewalPeriod(0).
+		WithExplicitClose(true)
 
-		// add indexes to collection
-		for name, idx := range addIdxKeys {
-			updateCollectionStmts = append(updateCollectionStmts, sql.NewCreateIndexStmt(collectionName, []string{name}, idx.IsUnique))
-		}
+	sqlTx, err := e.sqlEngine.NewTx(ctx, opts)
+	if err != nil {
+		return mayTranslateError(err)
+	}
+	defer sqlTx.Cancel()
+
+	table, err := getTableForCollection(sqlTx, collectionName)
+	if err != nil {
+		return err
+	}
 
+	currIDFieldName := docIDFieldName(table)
+
+	if idFieldName != "" && idFieldName != currIDFieldName {
 		_, _, err := e.sqlEngine.ExecPreparedStmts(
 			ctx,
 			sqlTx,
-			updateCollectionStmts,
+			[]sql.SQLStmt{sql.NewRenameColumnStmt(table.Name(), idFieldName, idFieldName)},
 			nil,
 		)
 		if err != nil {
-			return err
+			return mayTranslateError(err)
 		}
 	}
 
-	return sqlTx.Commit(ctx)
+	err = sqlTx.Commit(ctx)
+	return mayTranslateError(err)
 }
 
 // DeleteCollection deletes a collection.
 func (e *Engine) DeleteCollection(ctx context.Context, collectionName string) error {
-	// delete collection from catalog
-	_, ctxs, err := e.sqlEngine.ExecPreparedStmts(
+	opts := sql.DefaultTxOptions().
+		WithUnsafeMVCC(true).
+		WithSnapshotMustIncludeTxID(func(lastPrecommittedTxID uint64) uint64 { return 0 }).
+		WithSnapshotRenewalPeriod(0).
+		WithExplicitClose(true)
+
+	sqlTx, err := e.sqlEngine.NewTx(ctx, opts)
+	if err != nil {
+		return mayTranslateError(err)
+	}
+	defer sqlTx.Cancel()
+
+	_, _, err = e.sqlEngine.ExecPreparedStmts(
 		ctx,
 		nil,
```

Comment:
you are not passin in this sqlTx to ExecPreparedStmts, it would create a new transaction underneath